vscode-extension/client/src/extension.ts

```diff
@@ -12,6 +21,89 @@ import {
 
 let client: LanguageClient;
 
+interface LSPFileResponse {
+  content: string;
+}
+
+class SimpleRemoteFileProvider implements FileSystemProvider {
+  private _emitter = new EventEmitter<vscode.FileChangeEvent[]>();
+  readonly onDidChangeFile = this._emitter.event;
+
+  static fileContents = new Map<string, Uint8Array>();
+
+  async readFile(uri: Uri): Promise<Uint8Array> {
+    try {
+      // Forward the request to LSP and get the response
+      const response = await client.sendRequest<LSPFileResponse>(
+        "fileSystem/read",
+        {
+          uri: uri.toString(),
+        },
+      );
+
+      if (!response || typeof response.content !== "string") {
+        throw new Error("Invalid response from LSP server");
+      }
+
+      // Convert response content to Uint8Array and cache it
+      const content = new TextEncoder().encode(response.content);
+      SimpleRemoteFileProvider.fileContents.set(uri.toString(), content);
+      return content;

```

Comment:
Maybe we don't know but:
why do we need both of these lines? they seem to do the same thing.
Maybe the first line sets the content for any _opened_ documents, and the `return` saves the content for any future openings?

Reply:
I got the caching process to work. Now we only make LSP requests when reading new files